---
output: html_document
bibliography: "`r sharedbib::bib_path()`"
---

```{r include=FALSE}
library(sharedbib)
```


# Summary

[A short summary of the subject as a whole in historical context]




# Notes

[A nested summary of information derived from the below sources organized by topic]




# Articles


## @kumar2016advances

`r link_document('kumar2016advances')`

Describes the evolution of techniques used to scale phylogenetic branch length by historical time (aka time trees), grouping them into 4 "generations"

### Relevance

Background reading for TreeTime

### Context


### Notes

* "In the first generation approaches (1960s–1980s), a strict molecular clock was assumed to date divergences. In the second generation approaches (1990s), the equality of evolutionary rates between species was first tested and then a strict molecular clock applied to estimate divergence times. The third generation approaches (since 2000) account for differences in evolutionary rates across the tree by using a statistical model, obviating the need to assume a clock or to test the equality of evolutionary rates among species. Bayesian methods in the third generation require a specific or uniform prior on the speciation-process and enable the inclusion of uncertainty in clock calibrations. The fourth generation approaches (since 2012) allow rates to vary from branch to branch, but do not need prior selection of a statistical model to describe the rate variation or the specification of speciation model."
* "A strong correlation between the level of sequence similarity and fossil-based divergence times supported the use of first generation methods"
* "Both third and fourth generation methods have been found to perform well even when the sequence alignments are missing a substantial portion of the data (Filipski et al. 2014; Zheng and Wiens 2015)."



## @sagulenko2018treetime

`r link_document('sagulenko2018treetime')`

### Relevance

* Default program used for the `augur` `refine` and `ancestral` commands.

### Context

* Claimed to be a compromise between BEAST's accuracy and the speed of distance-based approches

### Notes

* I could not understand a lot of the algorithm details
* Composed of a set of python functions that can also be run from the command line
* O(n) time
* Designed to work with alignments of rapidly-evolving viruses sampled at different times
* "The main applications of TreeTime are ancestral state inference, evolutionary model inference, and time tree estimation."
* "TreeTime uses probabilistic models of evolution, allows inference of ancestral characters, and coalescent models. In TreeTime, every node of the tree can be given a strict or probabilistic date constraint."
* "TreeTime was tested predominantly on sequences from viruses with a pairwise identity above 90%. The iterative optimization procedures are not expected to be accurate for trees were many sites are saturated"

## @minh2020iq

`r link_document('minh2020iq')`

### Relevance

* Default program used for the `augur` `tree` command.

### Context

* Describes upgrades to `IQ-TREE` resulting in `IQ-TREE 2`

### Notes

* I could not understand a lot of it, since it gets into the details of evolutionary models and ML tree inference
* Can produce rooted trees
* Can enforce the monophyly of specified groups
* Can automatically run multiple times for difficult trees
* Can infer multiple individual locus/gene trees for use with coalescent analysis.
* More multithreading and cluster support
* Can avoid using too much RAM
* Can resume a failed run

## @nguyen2015iq

`r link_document('nguyen2015iq')`

### Relevance

* Default program used for the `augur` `tree` command.

### Context

* Claims to be a faster alternative to RAXML and PYHML, which they said were the most popular programs at the time.

### Notes

* Generally faster and more accurate than RAxML
* Uses a combination of standard hill-climbing approaches via nearest-neighbor interchange (NNI), a random NNI to escape local maxima, and a set of locally-optimal trees  



# To read

* Kumar, S., and Hedges, S. B. (2016) ‘Advances in Time
Estimation Methods for Molecular Data’, Molecular Biology and
Evolution, 33: 863.




# References

[Where the bibliography is automatically added]
